Warn (don't fail!) on spelling errors. Fix typos reported by codespell.

practicalswift opened the pull request:
* Check for common misspellings using `codespell`.
* Fix recently introduced typos reported by `codespell`.

practicalswift committed:
lint: Add spell check linter (codespell)

fanquake commented:
Travis failure:
```
33.63s$ test/lint/lint-all.sh
test/lint/lint-all.sh: line 21: test/lint/lint-spelling.sh: Permission denied
^---- failure generated from test/lint/lint-spelling.sh
```

ken2812221 commented on .travis.yml:
```diff
@@ -143,11 +143,11 @@ jobs:
       language: python
       python: '3.6'
       install:
+        - travis_retry sudo apt-get install --no-install-recommends --no-upgrade -qq codespell
```
Can use
```
addons:
  apt:
    packages:
    - codespell
```
to avoid sudo

practicalswift replied:
Oh that is much cleaner! Thanks!

laanwj commented:
Does this mean we need to manually add words that are not recognized to the spell checker?
If we do this, I think this certainly needs maintenance instructions.

Empact commented on test/lint/lint-spelling.sh:
```diff
@@ -0,0 +1,22 @@
+
+EXIT_CODE=0
+EXCLUDE_REGEXP="(src/example.cpp:.*example-false-positive-word) "
+SPELLING_ERRORS=$(codespell -d -q 3 $(git ls-files -- ":(exclude)src/leveldb/" ":(exclude)src/secp256k1/" ":(exclude)src/univalue/" ":(exclude)depends/" ":(exclude)doc/release-notes/" ":(exclude)src/qt/locale/" ":(exclude)test/functional/test_framework/__init__.py" ":(exclude)test/lint/lint-spelling.sh") | grep -vE "${EXCLUDE_REGEXP}")
```
nit: using the long form options here would be more self-documenting: `--disable-colors --quiet-level=`

practicalswift replied:
Good point! Now fixed.

Empact commented on src/rpc/net.cpp:
```diff
@@ -242,13 +242,13 @@ static UniValue disconnectnode(const JSONRPCRequest& request)
+            "\n" "Immediately disconnects from the specified peer node.\n"
+            "\n" "Strictly one out of 'address' and 'nodeid' can be provided to identify the node.\n"
+            "\n" "To disconnect by nodeid, either set 'address' to the empty string, or call using the named 'nodeid' argument only.\n"
+            "\n" "Arguments:\n"
```
Why are these changes necessary? There are plenty more leading `\n` in this file.

practicalswift replied:
The old Ubuntu packaged version didn't understand `\n` and thus flagged "nTo" as a typo. That seems to have been fixed in the more recent `pip` installed version.

Empact commented on .travis.yml:
```diff
@@ -147,6 +147,10 @@ jobs:
+      addons:
+        apt:
+          packages:
+          - codespell
```
`codespell` recommends installation via `pip`:
https://github.com/codespell-project/codespell

practicalswift replied:
Good point! Now fixed

Empact commented:
Concept ACK

MarcoFalke commented on src/qt/paymentserver.h:
```diff
@@ -77,7 +77,7 @@ class PaymentServer : public QObject
     // or, if that's not set, to use the system default root certificates.
     // If you pass in a store, you should not X509_STORE_free it: it will be
-    // freed either at exit or when another set of CAs are loaded.
+    // freed either at exit or when another set of CA:s are loaded.
```
why?

Empact replied:
Yeah I would add these to `test/lint/lint-spelling.ignore-words.txt` rather than use `:` to separate them.

MarcoFalke commented on src/qt/coincontroldialog.cpp:
```diff
@@ -185,7 +185,7 @@ void CoinControlDialog::buttonSelectAllClicked()
     ui->treeWidget->setEnabled(true);
     if (state == Qt::Unchecked)
-        coinControl()->UnSelectAll(); // just to be sure
+        coinControl()->DeSelectAll(); // just to be sure
```
Should probably be a scripted diff

laanwj commented:
I'm... not convinced about this.

I think running this once in a while is good, just to see if there's any really awkward misspellings, but running this in travis and failing the tests just because of a misspelled word in a comment seems overkill.

As I've said before, many times, linters should be added cautiously, there to prevent real problems resulting in bugs.

practicalswift committed:
Fix typos reported by codespell

practicalswift commented:
@laanwj The reason I wrote this linter is your comment in #9962 which made clear that your preference was to catch typos as review time so that they never get merged in the first place, and that this should be done without involving the maintainers.

From https://github.com/bitcoin/bitcoin/pull/9962#issuecomment-285332911:

> In principle it is good to fix comment typos, but my preference too would be to either catch them at the source (review the PR and find them) so that they never get merged in the first place or bunch them up into something like a "fix typos" pull every few weeks.

From https://github.com/bitcoin/bitcoin/pull/9962#issuecomment-285334709:

> […] This allows live cooperation on fixing typos without involving the maintainers at every step.

Also @fanquake requested an automated solution in #10419:

> Im curious. What tool are you using to find these? I assume it's being run in some automated way on the repository when new changes are merged? Why not merge the open pull requests instead, catch the typos/incorrect includes etc etc and then advise changes in the PRs while they are still open?

scravy commented:
Maybe it would make sense to have this run as a separate check (as in github check – https://developer.github.com/v3/checks/runs/ ). Especially with things like an AppVeyor build for MSVC (https://github.com/bitcoin/bitcoin/pull/13964#issuecomment-412852328, which already is a different check) it might be handy to have this split out of the other builds.

It would need to be hosted somewhere though I guess.

practicalswift commented:
@laanwj Can you clarify your position? :-)

Instead of exiting with a failure if new typos are introduced in a submitted PR we could perhaps just let `codespell` print the identified typos to the console as part of the linting process. Sounds like a good compromise?

practicalswift commented:
Now only warns in case of spelling errors in order to not Travis. Also rebased :-)

practicalswift commented:
The output can be seen on https://travis-ci.org/bitcoin/bitcoin/jobs/421142979:

```
…
$ set -o errexit; source .travis/lint_06_script.sh
Running script for: 52bc0698ca0b27cd24f6f31b335495d641fd58cf
sed -i 's/UnSelect/DeSelect/g' src/qt/coincontroldialog.cpp src/qt/sendcoinsdialog.cpp src/wallet/coincontrol.h
OK
src/crypto/ctaes in HEAD currently refers to tree 1b6c31139a71f80245c09597c343936a8e41d021
src/crypto/ctaes in HEAD was last updated in commit 8501bedd7508ac514385806e191aec21ee978891 (tree 1b6c31139a71f80245c09597c343936a8e41d021)
subtree commit 003a4acfc273932ab8c2e276cde3b4f3541012dd unavailable: cannot compare
src/secp256k1 in HEAD currently refers to tree af619602c243e0d8fbd5934f375faa4aedb4ca6e
src/secp256k1 in HEAD was last updated in commit fd86f998fcfd25d823d67a2920814e22445655f9 (tree af619602c243e0d8fbd5934f375faa4aedb4ca6e)
subtree commit 0b7024185045a49a1a6a4c5615bf31c94f63d9c4 unavailable: cannot compare
src/univalue in HEAD currently refers to tree a2dbaf30021c0a31eecc213cd84d600273d333b6
src/univalue in HEAD was last updated in commit a570098021be6a7b9f4589300ea655ae4633628e (tree a2dbaf30021c0a31eecc213cd84d600273d333b6)
subtree commit 51d3ab34ba2857f0d03dc07250cb4a2b5e712e67 unavailable: cannot compare
src/leveldb in HEAD currently refers to tree f8cdc77167be86194f98d0412baf4b89d3d5abe6
src/leveldb in HEAD was last updated in commit ec749b1bcdf2483b642fb51d635800e272c68ba6 (tree f8cdc77167be86194f98d0412baf4b89d3d5abe6)
subtree commit 524b7e36a8e3bce6fcbcd1b5df09024283f325ba unavailable: cannot compare
Args used        : 181
Args documented  : 186
Args undocumented: 0
set()
Args unknown     : 5
{'-zmqpubhashtx', '-zmqpubhashblock', '-zmqpubrawtx', '-promiscuousmempoolflags', '-zmqpubrawblock'}
* Checking consistency between dispatch tables and vRPCConvertParams
src/streams.h:532: signficant  ==> significant
^ Warning: codespell identified likely spelling errors. Any false positives? Add them to the list of ignored words in test/lint/lint-spelling.ignore-words.txt
The command "set -o errexit; source .travis/lint_06_script.sh" exited with 0.
…
```

Pretty sweet :-)

practicalswift commented:
Rebased and fixed the typos that `codespell` found in the newly introduced documentation.

`codespell` catched the "account" typo first reported by @promag in PR #14023 post-merge: https://github.com/bitcoin/bitcoin/pull/14023#discussion_r213136227

jnewbery commented on src/dbwrapper.cpp:
```diff
@@ -78,7 +78,7 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
     // also OK, because up to that amount LevelDB will use an mmap
     // implementation that does not use extra file descriptors (the fds are
-    // closed after being mmaped).
+    // closed after being mmap:ed).
```
I've never seen a colon used in this way. Why not mmap'ed?

jnewbery commented on src/qt/coincontroldialog.cpp:
```diff
@@ -185,7 +185,7 @@ void CoinControlDialog::buttonSelectAllClicked()
     ui->treeWidget->setEnabled(true);
     if (state == Qt::Unchecked)
-        coinControl()->UnSelectAll(); // just to be sure
+        coinControl()->DeSelectAll(); // just to be sure
```
Is unselect really not ok? https://en.oxforddictionaries.com/definition/unselect.

practicalswift replied:
@jnewbery I found this: https://english.stackexchange.com/questions/18465/unselect-or-deselect

I'm happy to revert this specific change if it is controversial :-)

jnewbery reviewed (commented):
Apart from the unselect->deselect change, I think the spelling corrections are fine.

Concept ACK on the linting, although I haven't reviewed the code. I'd lean toward having the linter fail on spelling errors rather than just warn (which will no doubt be ignored).

practicalswift commented:
@jnewbery Thanks for the encouraging review! PR updated – please re-review :-)

If this PR is merged I'll look if there is some smart way to use Travis folding or similar to make the spelling warnings more prominent in the Travis interface.

laanwj commented:
>  I'd lean toward having the linter fail on spelling errors rather than just warn (which will no doubt be ignored).

I still strongly disagree with this, I don't think we should pile up linters for cosmetic issues.

The original idea of them was to flag serious issues that can result in bugs.

Failing the tests on some supposed spelling error in a comment really, goes too far.

practicalswift commented:
@laanwj What about enabling it on a warning level (as the PR is currently formulated)? Could that be a good compromise? :-)

jnewbery commented:
> I still strongly disagree with this

ok, I won't rehash the debate that we've all had in countless other PRs already :)

My point was merely that warnings tend to get ignored, so in general I think they're less useful than errors. I don't have any strong opinion on this PR though.

practicalswift commented:
@jnewbery When thinking about it again I agree regarding "deselect" vs "unselect". Now added it to the ignore list instead and removed that commit. Please re-review :-)

Sjors commented:
Concept ACK for warnings. Maybe Drahbot can point out typos. In addition we could fix the ones we missed as part of the release process.

utACK 4a98785 (though maybe squash the first and last commit)

practicalswift commented:
@Sjors Squashed. Please re-review :-)

Sjors commented:
utACK b5ee29e460dc36066bac68e041073100c0eca20e

fanquake commented on src/qt/coincontroldialog.cpp:
```diff
@@ -441,7 +441,7 @@ void CoinControlDialog::updateLabels(WalletModel *model, QDialog* dialog)
         if (out.depth_in_main_chain < 0) continue;
 
-        // unselect already spent, very unlikely scenario, this could happen
+        // deselect already spent, very unlikely scenario, this could happen
```
If unselect is in the ignore list, why is this being changed?

practicalswift replied:
Good point. Fixed!

practicalswift commented:
@Sjors @fanquake Updated. Please re-review :-)

MarcoFalke commented:
utACK f8a81f73ac4fee82488806c2b384b36e8ee2acdf

Empact commented on test/lint/lint-spelling.ignore-words.txt:
```diff
@@ -0,0 +1,6 @@
+cas
+hights
```
I don't find `hights` in current master

practicalswift replied:
@Empact 

```
$ git grep -i hights
contrib/linearize/linearize-data.py:        self.highTS = 1408893517 - 315360000
contrib/linearize/linearize-data.py:                os.utime(self.outFname, (int(time.time()), self.highTS))
contrib/linearize/linearize-data.py:                    os.utime(self.outFname, (int(time.time()), self.highTS))
contrib/linearize/linearize-data.py:        if blkTS > self.highTS:
contrib/linearize/linearize-data.py:            self.highTS = blkTS
```

:-)

Empact commented:
utACK ada3562

practicalswift commented:
@laanwj Oh, I wasn't aware of the policy change introduced in #14179 so I don't have anything meaningful to add about that. Setting the discussion about that merge decision aside:

Do you have a link to the Travis failure? It would be interesting to see if it was a false positive or a false positive that caused the Travis failure.

luke-jr commented:
It came up when someone tried to name a mutex `mut`. Seems unreasonable to demand variables have English word names...

laanwj commented:
yes it came up on IRC @karel-3d
```
08:29 < karelb> I am trying to run my fork of bitcoind on travis.... and travis stops with this
08:29 < karelb> src/threadinterrupt.cpp:25: mut  ==> must, mutt, moot
08:29 < karelb> Warning: codespell identified likely spelling errors
08:29 < karelb> failure generated from test/lint/lint-spelling.sh
08:29 < karelb> .....ummmmm, ok?
08:30 < karelb> I did no change in threadinterrupt, and that "mut" is just some mutex
08:39 < karelb> well whatever, one commit that renamed "mut" to "mutex" fixed that, I just wonder why I needed to do that...
08:40 < gwillen> karelb: your travis instance is stupid
08:40 < gwillen> if you make it less stupid you will have fewer problems
08:40 < gwillen> it is trying to spell-correct your code, and doing it moronically
08:40 < karelb> :D
08:40 < gwillen> and seems to be set to warnings-as-errors.
08:40 < karelb> I use travis-ci.org
08:41 < gwillen> well, apparently they are stupid
08:41 < gwillen> but I imagine there is an option not to make codespell warnings errors
08:41 < karelb> the same config as bitcoin core
08:41 < gwillen> which they absolutely should not be
08:41 < gwillen> oh, hm
08:41 < gwillen> I mean, seemingly not
08:42 < karelb> well let's see what happens when I make a PR, if that `mut` stuff is still there
10:49 < wumpus> karelb: lol that's stupid
10:50 < wumpus> please don't tell me that stupid spelling check is mandatory in travis now
```

scravy commented:
@practicalswift Which versions of codespell?

practicalswift commented:
@scravy 

```
$ pip3 install codespell --upgrade
Requirement already up-to-date: codespell in /usr/local/lib/python3.6/dist-packages
$ codespell --version
1.14.0
```

practicalswift commented:
Fixed in #14495 :-)

practicalswift commented:
@scravy Seems like this was introduced somewhere between `1.13.0` and `1.14.0`. We're pinning to `1.13.0` in Travis so I don't know how `1.13.0` got installed in @karel-3d:s Travis setup. @karel-3d, do you know? :-)

karel-3d commented:
I cloned the repo, did some experimental changes, and used Travis-ci.org.

The build is here

https://travis-ci.org/karel-3d/bitcoin/builds/442019337

It's branched from this commit, 18 days old. (Maybe it will go away if I rebase on current master?)

https://github.com/bitcoin/bitcoin/commit/0809e68a9084630ff4d11ec0503a3d0ab52bc6d7

karel-3d commented:
Oh yeah. It seems I have branched the master *before* this commit

https://github.com/bitcoin/bitcoin/commit/d10f2cd7d8506aec8b406c346b35d3c31f0f37df

Which means that if I rebase to master, it would indeed go away.

practicalswift commented:
@karel-3d Thanks for the clarification! That explains the mystery :-)
